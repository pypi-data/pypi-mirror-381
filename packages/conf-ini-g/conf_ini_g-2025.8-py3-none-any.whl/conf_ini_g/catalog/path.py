"""
Copyright CNRS/Inria/UniCA
Contributor(s): Eric Debreuve (eric.debreuve@cnrs.fr) since 2021
SEE COPYRIGHT NOTICE BELOW
"""

import dataclasses as d
import typing as h
from pathlib import Path as pl_path_t

from babelwidget.api.backend import backend_t
from babelwidget.api.dialog import path_chooser_t
from babelwidget.api.widget import base_h as library_wgt_h
from babelwidget.api.widget import label_p as label_wgt_h
from babelwidget.api.widget import text_line_p as text_line_wgt_h
from conf_ini_g.constant.window import PATH_SELECTOR_WIDTH
from conf_ini_g.interface.window.parameter.value import value_wgt_a
from value_factory.api.catalog import path_kind_e, path_purpose_e, path_t
from value_factory.api.type import hint_t


@d.dataclass(repr=False, eq=False)
class path_wgt_t(value_wgt_a):
    """
    Cannot use slots (weak reference issue).
    """

    PostAssignmentFunction: h.Callable[[pl_path_t], None] | None
    library_wgt: library_wgt_h
    editable: bool
    type_: path_kind_e | None = d.field(init=False, default=None)
    path: text_line_wgt_h | label_wgt_h | None = d.field(init=False, default=None)
    _path_chooser: path_chooser_t | None = d.field(init=False, default=None)

    @classmethod
    def New(
        cls,
        stripe: hint_t,
        backend: backend_t,
        /,
        *,
        editable: bool = True,
        PostAssignmentFunction: h.Callable[[pl_path_t], None] | None = None,
    ) -> h.Self:
        """
        If stripe does not contain the necessary details, the target type is set to any and considered as input, and
        the selection button label ends with an exclamation point.
        """
        if editable:
            path = backend.text_line_t()
            messenger = path
        else:
            path = backend.label_t()
            messenger = None
        output = cls(
            messenger,
            "textChanged",
            backend,
            PostAssignmentFunction=PostAssignmentFunction,
            library_wgt=backend.base_t(),
            editable=editable,
        )
        output.path = path

        annotation = stripe.FirstAnnotationWithType(path_t)
        if annotation is None:
            path_purpose = path_purpose_e.input
            path_kind = path_kind_e.any
        else:
            path_purpose = annotation.purpose
            path_kind = annotation.kind

        output.type_ = path_kind
        output._path_chooser = path_chooser_t.New(
            path_purpose.name, path_kind.name, backend
        )

        if path_kind is path_kind_e.document:
            selector_label = "üóã"
        elif path_kind is path_kind_e.folder:
            selector_label = "üìÇ"
        else:
            selector_label = "üìÇüóã"
        if path_purpose is path_purpose_e.input:
            selector_color = "green"
        elif path_purpose is path_purpose_e.output:
            selector_color = "red"
        else:
            selector_color = "blue"
        path_selector = backend.button_t(selector_label, parent=output.library_wgt)
        path_selector.SetFunction(output.SelectDocument)

        path_selector.setStyleSheet(f"color: {selector_color};")
        path_selector.setFixedWidth(PATH_SELECTOR_WIDTH)

        layout = backend.layout_h_t()
        layout.setContentsMargins(0, 0, 0, 0)
        layout.addWidget(path)
        layout.addWidget(path_selector)
        output.library_wgt.setLayout(layout)

        return output

    def Assign(self, value: pl_path_t, _: h.Any, /) -> None:
        """"""
        value = _ValidStrValue(value, self.editable)
        self.path.setText(value)

    def Text(self) -> str | None:
        """
        /!\\ pathlib.Path("") == pathlib.Path(".").
        """
        return self.path.Text()

    def SelectDocument(self) -> None:
        """"""
        current_path = self.Text()
        current_doc = pl_path_t(current_path).resolve()

        # Used to use self.backend_for_selection.
        self._path_chooser.SetFolder(current_doc.parent)
        selection = self._path_chooser.NewSelected()
        if selection is None:
            return

        self.Assign(selection, None)
        # Put post-assignment call here instead of in the Assign method in case Assign
        # is also called at initialization time, in New, one day. Indeed, the
        # post-assignment task should have been done already, or will be done, when
        # instantiating the interface.
        if self.PostAssignmentFunction is not None:
            self.PostAssignmentFunction(selection)


def _ValidStrValue(value: pl_path_t, editable: bool, /) -> str:
    """"""
    if not editable:
        try:
            value = value.resolve(strict=True).relative_to(
                pl_path_t.home().resolve(strict=True)
            )
        except ValueError:
            # On Linux, this happens when home is a bind mount.
            pass

    return str(value)


"""
COPYRIGHT NOTICE

This software is governed by the CeCILL  license under French law and
abiding by the rules of distribution of free software.  You can  use,
modify and/ or redistribute the software under the terms of the CeCILL
license as circulated by CEA, CNRS and INRIA at the following URL
"http://www.cecill.info".

As a counterpart to the access to the source code and  rights to copy,
modify and redistribute granted by the license, users are provided only
with a limited warranty  and the software's author,  the holder of the
economic rights,  and the successive licensors  have only  limited
liability.

In this respect, the user's attention is drawn to the risks associated
with loading,  using,  modifying and/or developing or reproducing the
software by the user in light of its specific status of free software,
that may mean  that it is complicated to manipulate,  and  that  also
therefore means  that it is reserved for developers  and  experienced
professionals having in-depth computer knowledge. Users are therefore
encouraged to load and test the software's suitability as regards their
requirements in conditions enabling the security of their systems and/or
data to be ensured and,  more generally, to use and operate it in the
same conditions as regards security.

The fact that you are presently reading this means that you have had
knowledge of the CeCILL license and that you accept its terms.

SEE LICENCE NOTICE: file README-LICENCE-utf8.txt at project source root.

This software is being developed by Eric Debreuve, a CNRS employee and
member of team Morpheme.
Team Morpheme is a joint team between Inria, CNRS, and UniCA.
It is hosted by the Centre Inria d'Universit√© C√¥te d'Azur, Laboratory
I3S, and Laboratory iBV.

CNRS: https://www.cnrs.fr/index.php/en
Inria: https://www.inria.fr/en/
UniCA: https://univ-cotedazur.eu/
Centre Inria d'Universit√© C√¥te d'Azur: https://www.inria.fr/en/centre/sophia/
I3S: https://www.i3s.unice.fr/en/
iBV: http://ibv.unice.fr/
Team Morpheme: https://team.inria.fr/morpheme/
"""
