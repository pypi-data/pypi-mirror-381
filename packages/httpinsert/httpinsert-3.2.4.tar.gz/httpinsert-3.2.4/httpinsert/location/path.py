from urllib.parse import urlunparse, quote,urlparse
from httpinsert.location import Location
from httpinsert.insertion_points import InsertionPoint

class Path(Location):

    def find_insertion_points(self,request):
        insertion_points=[]
        parsed_url = urlparse(request.url)
        path = parsed_url.path
        segments = path.strip("/").split("/")
        for idx, segment in enumerate(segments):
            insertion_points.append(InsertionPoint(self,"path", str(idx), segment,default=False))
        if path != "/" and path[-1] == "/":
            insertion_points.append(InsertionPoint(self,"path",str(idx+1),"",default=False,new_param=True))
        return insertion_points


    def insert_payload(self,request,insertion_point,payload,default_encoding):
        if default_encoding is True:
            if isinstance(payload,str):
                payload=quote(payload) # Encode the payload
            else:
                payload = [quote(i) for i in payload]
        parsed_url = urlparse(request.url)
        path = parsed_url.path
        segments = path.strip("/").split("/")
        if int(insertion_point.key) == len(segments):
            path+=payload
        else:
            segments[int(insertion_point.key)] = payload
            path = quote("/" + "/".join(segments))
        url = urlunparse(parsed_url._replace(path=path))
        request.url=url
        return request,request.url


Path()
