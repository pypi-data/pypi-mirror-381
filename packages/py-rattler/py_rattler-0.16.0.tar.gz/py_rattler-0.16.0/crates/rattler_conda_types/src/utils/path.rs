use itertools::Itertools;

/// Returns true if the specified string is considered to be an absolute path
pub(crate) fn is_absolute_path(path: &str) -> bool {
    if path.contains("://") {
        return false;
    }

    // Check if the path starts with a common absolute path prefix
    if path.starts_with('/') || path.starts_with("\\\\") {
        return true;
    }

    // A drive letter followed by a colon and a (backward or forward) slash
    matches!(path.chars().take(3).collect_tuple(),
        Some((letter, ':', '/' | '\\')) if letter.is_alphabetic())
}

/// Returns true if the specified string is considered to be a path
pub(crate) fn is_path(path: &str) -> bool {
    if path.contains("://") {
        return false;
    }

    // Check if the path starts with a common path prefix
    if path.starts_with("./")
        || path.starts_with("..")
        || path.starts_with("~/")
        || path.starts_with('/')
        || path.starts_with("\\\\")
        || path.starts_with("//")
    {
        return true;
    }

    // A drive letter followed by a colon and a (backward or forward) slash
    matches!(path.chars().take(3).collect_tuple(),
        Some((letter, ':', '/' | '\\')) if letter.is_alphabetic())
}

mod tests {
    #[test]
    fn test_is_absolute_path() {
        use super::is_absolute_path;
        assert!(is_absolute_path("/foo"));
        assert!(is_absolute_path("/C:/foo"));
        assert!(is_absolute_path("C:/foo"));
        assert!(is_absolute_path("\\\\foo"));
        assert!(is_absolute_path("\\\\server\\foo"));

        assert!(!is_absolute_path("conda-forge/label/rust_dev"));
        assert!(!is_absolute_path("~/foo"));
        assert!(!is_absolute_path("./foo"));
        assert!(!is_absolute_path("../foo"));
        assert!(!is_absolute_path("foo"));
        assert!(!is_absolute_path("~\\foo"));
    }

    #[test]
    fn test_is_path() {
        use super::is_path;
        assert!(is_path("/foo"));
        assert!(is_path("/C:/foo"));
        assert!(is_path("C:/foo"));
        assert!(is_path("\\\\foo"));
        assert!(is_path("\\\\server\\foo"));

        assert!(is_path("./conda-forge/label/rust_dev"));
        assert!(is_path("~/foo"));
        assert!(is_path("./foo"));
        assert!(is_path("../foo"));

        assert!(!is_path("~\\foo"));
    }
}
