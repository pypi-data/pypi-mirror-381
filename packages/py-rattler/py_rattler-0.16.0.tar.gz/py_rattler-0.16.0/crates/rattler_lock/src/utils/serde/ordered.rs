use serde::{Deserialize, Deserializer, Serialize, Serializer};
use serde_with::__private__::{DeserializeAsWrap, SerializeAsWrap};
use serde_with::{DeserializeAs, SerializeAs};
use std::collections::HashSet;
use std::hash::{BuildHasher, Hash};
use std::marker::PhantomData;

/// Used with `serde_with` to serialize a collection as a sorted collection.
#[derive(Default)]
pub(crate) struct Ordered<T>(PhantomData<T>);

impl<'de, T: Eq + Hash, S: BuildHasher + Default, TAs> DeserializeAs<'de, HashSet<T, S>>
    for Ordered<TAs>
where
    TAs: DeserializeAs<'de, T>,
{
    fn deserialize_as<D>(deserializer: D) -> Result<HashSet<T, S>, D::Error>
    where
        D: Deserializer<'de>,
    {
        let content =
            DeserializeAsWrap::<Vec<T>, Vec<TAs>>::deserialize(deserializer)?.into_inner();
        Ok(content.into_iter().collect())
    }
}

impl<T: Ord, HS, TAs: SerializeAs<T>> SerializeAs<HashSet<T, HS>> for Ordered<TAs> {
    fn serialize_as<S>(source: &HashSet<T, HS>, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let mut elements: Vec<_> = source.iter().collect();
        elements.sort();
        SerializeAsWrap::<Vec<&T>, Vec<&TAs>>::new(&elements).serialize(serializer)
    }
}

impl<'de, T: Ord, TAs> DeserializeAs<'de, Vec<T>> for Ordered<TAs>
where
    TAs: DeserializeAs<'de, T>,
{
    fn deserialize_as<D>(deserializer: D) -> Result<Vec<T>, D::Error>
    where
        D: Deserializer<'de>,
    {
        let mut content =
            DeserializeAsWrap::<Vec<T>, Vec<TAs>>::deserialize(deserializer)?.into_inner();
        content.sort();
        Ok(content)
    }
}

impl<T: Ord, TAs: SerializeAs<T>> SerializeAs<Vec<T>> for Ordered<TAs> {
    fn serialize_as<S>(source: &Vec<T>, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let mut elements: Vec<_> = source.iter().collect();
        elements.sort();
        SerializeAsWrap::<Vec<&T>, Vec<&TAs>>::new(&elements).serialize(serializer)
    }
}
