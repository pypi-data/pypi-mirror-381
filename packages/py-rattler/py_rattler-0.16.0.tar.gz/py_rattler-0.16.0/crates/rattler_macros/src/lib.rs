//! Some macros for the Rattler project.

#![deny(missing_docs)]

use proc_macro::TokenStream;
use quote::quote_spanned;
use syn::{parse_macro_input, Data, DeriveInput, Fields, FieldsNamed, Ident};

/// Macro for enforcing alphabetical order on Structs and Enums.
///
/// This macro will not automatically sort it for you; rather, it will fail to compile if the
/// fields are not defined alphabetically.
#[proc_macro_attribute]
pub fn sorted(_attr: TokenStream, item: TokenStream) -> TokenStream {
    let out = item.clone();
    let input = parse_macro_input!(item as DeriveInput);

    let name = &input.ident;

    let check_sorted = match &input.data {
        Data::Struct(data_struct) => match &data_struct.fields {
            Fields::Named(fields) => check_fields_sorted(name, fields),
            _ => panic!("This macro only supports structs with named fields."),
        },
        Data::Enum(data_enum) => {
            let variants: Vec<&Ident> = data_enum.variants.iter().map(|v| &v.ident).collect();
            check_identifiers_sorted(name, &variants)
        }
        Data::Union(_) => panic!("This macro only supports structs and enums."),
    };

    if check_sorted.is_err() {
        check_sorted.err().unwrap()
    } else {
        out
    }
}

fn check_fields_sorted(outer_ident: &Ident, fields: &FieldsNamed) -> Result<(), TokenStream> {
    let mut prev_field: Option<&Ident> = None;
    for field in &fields.named {
        let current_field = field.ident.as_ref().unwrap();
        if let Some(prev) = prev_field {
            if *current_field < *prev {
                let error = format!(
                    "The field {current_field} must be sorted before {prev} in struct {outer_ident}.",
                );
                let tokens = quote_spanned! {current_field.span() =>
                    compile_error!(#error);
                };
                return Err(TokenStream::from(tokens));
            }
        }
        prev_field = Some(current_field);
    }
    Ok(())
}

fn check_identifiers_sorted(outer_ident: &Ident, idents: &[&Ident]) -> Result<(), TokenStream> {
    let mut prev_ident: Option<&Ident> = None;
    for ident in idents {
        if let Some(prev) = prev_ident {
            if *ident < prev {
                let error = format!(
                    "The field {ident} must be sorted before {prev} in enum {outer_ident}.",
                );
                let tokens = quote_spanned! {ident.span() =>
                    compile_error!(#error);
                };
                return Err(TokenStream::from(tokens));
            }
        }
        prev_ident = Some(ident);
    }
    Ok(())
}
