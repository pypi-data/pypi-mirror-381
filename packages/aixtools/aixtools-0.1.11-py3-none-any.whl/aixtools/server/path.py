"""
Workspace path handling for user sessions.
"""

from pathlib import Path, PurePath, PurePosixPath

from fastmcp import Context

from ..utils.config import DATA_DIR
from .utils import get_session_id_tuple

WORKSPACES_ROOT_DIR = DATA_DIR / "workspaces"  # Path on the host where workspaces are stored
CONTAINER_WORKSPACE_PATH = PurePosixPath("/workspace")  # Path inside the sandbox container where workspace is mounted


def get_workspace_path(service_name: str = None, *, in_sandbox: bool = False, ctx: Context | tuple = None) -> PurePath:
    """
    Get the workspace path for a specific service (e.g. MCP server).
    If `service_name` is None, then the path to entire workspace folder (as mounted to a sandbox) is returned.
    If `in_sandbox` is True, it returns a path in sandbox, e.g.: `/workspace/mcp_repl`.
    If `in_sandbox` is False, it returns the path based on user and session IDs in the format:
    `<DATA_DIR>/workspaces/<user_id>/<session_id>/<service_name>`, where `DATA_DIR` should come from
    the environment variables, e.g.:
    `/data/workspaces/foo-user/bar-session/mcp_repl`.
    The `ctx` is used to get user and session IDs tuple. It can be passed directly or via HTTP headers from `Context`.
    If `ctx` is None, the current FastMCP request HTTP headers are used.

    Args:
        ctx: The FastMCP context, which contains the user session.
        service_name: The name of the service (e.g. "mcp_server").
        in_sandbox: If True, use a sandbox path; otherwise, use user/session-based path.

    Returns: The workspace path as a PurePath object.
    """
    if in_sandbox:
        path = CONTAINER_WORKSPACE_PATH
    else:
        user_id, session_id = ctx if isinstance(ctx, tuple) else get_session_id_tuple(ctx)
        path = WORKSPACES_ROOT_DIR / user_id / session_id
    if service_name:
        path = path / service_name
    return path


def container_to_host_path(path: PurePosixPath, *, ctx: Context | tuple = None) -> Path | None:
    """
    Convert a path in a sandbox container to a host path

    Args:
        container_path: Path inside the container (must be a subdir of CONTAINER_WORKSPACE_PATH).
        user_id: ID of the user.
        session_id: ID of the session.

    Returns:
        Path to the file on the host, or None if the conversion fails.
    """
    old_root = CONTAINER_WORKSPACE_PATH
    new_root = get_workspace_path(ctx=ctx)
    try:
        return new_root / PurePosixPath(path).relative_to(old_root)
    except ValueError as e:
        raise ValueError(f"Container path must be a subdir of '{old_root}', got '{path}' instead") from e


def host_to_container_path(path: Path, *, ctx: Context | tuple = None) -> PurePosixPath:
    """Convert a host path to a path in a sandbox container."""
    old_root = get_workspace_path(ctx=ctx)
    new_root = CONTAINER_WORKSPACE_PATH
    try:
        return new_root / Path(path).relative_to(old_root)
    except ValueError as exc:
        raise ValueError(f"Host path must be a subdir of '{old_root}', got '{path}' instead") from exc
