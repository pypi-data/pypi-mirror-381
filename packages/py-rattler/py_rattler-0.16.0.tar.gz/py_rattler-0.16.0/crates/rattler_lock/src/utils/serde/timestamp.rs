use chrono::{DateTime, Utc};
use serde::de::Error;
use serde::{Deserialize, Deserializer, Serialize, Serializer};
use serde_with::{DeserializeAs, SerializeAs};

pub(crate) struct Timestamp;

impl<'de> DeserializeAs<'de, chrono::DateTime<chrono::Utc>> for Timestamp {
    fn deserialize_as<D>(deserializer: D) -> Result<chrono::DateTime<chrono::Utc>, D::Error>
    where
        D: Deserializer<'de>,
    {
        let timestamp = i64::deserialize(deserializer)?;

        // Convert from milliseconds to seconds
        let microseconds = if timestamp > 253_402_300_799 {
            timestamp * 1_000
        } else {
            timestamp * 1_000_000
        };

        // Convert the timestamp to a UTC timestamp
        chrono::DateTime::from_timestamp_micros(microseconds)
            .ok_or_else(|| D::Error::custom("got invalid timestamp, timestamp out of range"))
    }
}

impl SerializeAs<chrono::DateTime<chrono::Utc>> for Timestamp {
    fn serialize_as<S>(source: &DateTime<Utc>, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        // Convert the date to a timestamp
        let timestamp: i64 = source.timestamp_millis();

        // Determine the precision of the timestamp.
        let timestamp = if timestamp % 1000 == 0 {
            timestamp / 1000
        } else {
            timestamp
        };

        // Serialize the timestamp
        timestamp.serialize(serializer)
    }
}
