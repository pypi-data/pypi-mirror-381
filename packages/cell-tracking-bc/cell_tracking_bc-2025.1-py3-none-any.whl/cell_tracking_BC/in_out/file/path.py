# Copyright CNRS/Inria/UCA
# Contributor(s): Eric Debreuve (since 2021)
#
# eric.debreuve@cnrs.fr
#
# This software is governed by the CeCILL  license under French law and
# abiding by the rules of distribution of free software.  You can  use,
# modify and/ or redistribute the software under the terms of the CeCILL
# license as circulated by CEA, CNRS and INRIA at the following URL
# "http://www.cecill.info".
#
# As a counterpart to the access to the source code and  rights to copy,
# modify and redistribute granted by the license, users are provided only
# with a limited warranty  and the software's author,  the holder of the
# economic rights,  and the successive licensors  have only  limited
# liability.
#
# In this respect, the user's attention is drawn to the risks associated
# with loading,  using,  modifying and/or developing or reproducing the
# software by the user in light of its specific status of free software,
# that may mean  that it is complicated to manipulate,  and  that  also
# therefore means  that it is reserved for developers  and  experienced
# professionals having in-depth computer knowledge. Users are therefore
# encouraged to load and test the software's suitability as regards their
# requirements in conditions enabling the security of their systems and/or
# data to be ensured and,  more generally, to use and operate it in the
# same conditions as regards security.
#
# The fact that you are presently reading this means that you have had
# knowledge of the CeCILL license and that you accept its terms.

import datetime as dttm
import sys as sstm
from typing import Union

from cell_tracking_BC.standard.number import MAX_INT
from cell_tracking_BC.standard.path import path_h, path_t
from logger_36 import L

# The following lists are meant to be safe enough, not to serve as references
PATH_ILLEGAL_CHARACTERS_LIN = r"/"
PATH_ILLEGAL_CHARACTERS_OSX = r":"
PATH_ILLEGAL_CHARACTERS_WIN = r'|/\<>:?*"'

REPLACEMENT_CHARACTER = "_"
VERSION_SEPARATOR = "-"

PATH_ILLEGAL_CHARACTERS = "".join(
    set(
        PATH_ILLEGAL_CHARACTERS_LIN
        + PATH_ILLEGAL_CHARACTERS_OSX
        + PATH_ILLEGAL_CHARACTERS_WIN
    )
)
if (REPLACEMENT_CHARACTER in PATH_ILLEGAL_CHARACTERS) or (
    VERSION_SEPARATOR in PATH_ILLEGAL_CHARACTERS
):
    raise ValueError(
        f'The character "{REPLACEMENT_CHARACTER}" or "{VERSION_SEPARATOR}" is an illegal path character'
    )


def StorageFolderForMain(
    base_folder: path_h,
    should_structure_folder: bool,
    sequence_name: str,
    first_frame: int,
    last_frame: int | None,
    /,
) -> path_t:
    """
    Call with main_py_module equal to __file__
    """
    if isinstance(base_folder, str):
        base_folder = path_t(base_folder)

    if should_structure_folder:
        output = ReplacePathIllegalCharacters(base_folder)
    else:
        output = base_folder
    if not output.exists():
        # exist_ok: added for thread safety (could use only mkdir then...)
        output.mkdir(exist_ok=True)
    elif not output.is_dir():
        L.error(f"{output}: Not a suitable storage folder")
        sstm.exit(-1)

    if not should_structure_folder:
        return output

    first_is_first = first_frame == 0
    last_is_last = (last_frame is None) or (last_frame == MAX_INT)
    if first_is_first and last_is_last:
        folder_name = sequence_name
    else:
        if first_is_first:
            from_frame = "FromStart"
        else:
            from_frame = f"From{first_frame}"
        if last_is_last:
            to_frame = "ToEnd"
        else:
            to_frame = f"To{last_frame}"
        folder_name = f"{sequence_name}-{from_frame}-{to_frame}"

    output /= ReplacePathIllegalCharacters(folder_name)
    if not output.exists():
        # exist_ok: added for thread safety (could use only mkdir then...)
        output.mkdir(exist_ok=True)
    elif not output.is_dir():
        L.error(f"{output}: Not a suitable storage folder")
        sstm.exit(-1)

    output /= TimeStamp()
    if output.exists():
        L.error(f"{output}: Existing date-based storage folder; Exiting")
        sstm.exit(-1)
    output.mkdir()

    return output


def TimeStamp() -> str:
    """"""
    return (
        dttm.datetime.now()
        .isoformat(timespec="milliseconds")
        .replace(".", "-")
        .replace(":", "-")
    )


def ReplacePathIllegalCharacters(
    path: Union[str, path_t], /, *, replacement: str = REPLACEMENT_CHARACTER
) -> Union[str, path_t]:
    """"""
    translations = str.maketrans(
        PATH_ILLEGAL_CHARACTERS, PATH_ILLEGAL_CHARACTERS.__len__() * replacement
    )
    if isinstance(path, str):
        as_path = path_t(path)
    else:
        as_path = path
    parts = as_path.parts
    if parts[0] == as_path.drive + as_path.root:
        output = [parts[0]]
        p_idx = 1
    else:
        output = []
        p_idx = 0
    for part in parts[p_idx:]:
        output.append(part.translate(translations))
    output = path_t(*output)

    if isinstance(path, str):
        return str(output)

    return output
