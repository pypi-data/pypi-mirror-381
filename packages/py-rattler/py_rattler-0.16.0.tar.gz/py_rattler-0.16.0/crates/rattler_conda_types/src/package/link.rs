use std::path::Path;

use rattler_macros::sorted;
use serde::{Deserialize, Serialize};

use super::{EntryPoint, PackageFile};

/// Describes python noarch specific entry points
#[derive(Serialize, Clone, Debug, Deserialize)]
pub struct PythonEntryPoints {
    /// A list of commands that should execute certain python commands.
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub entry_points: Vec<EntryPoint>,
}

/// Links for specific types of noarch packages.
#[derive(Serialize, Clone, Debug, Deserialize)]
#[serde(tag = "type", rename_all = "lowercase")]
pub enum NoArchLinks {
    /// Python noarch specific entry points.
    Python(PythonEntryPoints),

    /// Generic variant (doesn't have any special entry points)
    Generic,
}

/// A representation of the `link.json` file found in noarch package archives.
///
/// The `link.json` file contains information about entrypoints that need to be installed for the package.
#[sorted]
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct LinkJson {
    /// Links for specific noarch packages
    pub noarch: NoArchLinks,

    /// The version of the package metadata file
    pub package_metadata_version: u64,
}

impl PackageFile for LinkJson {
    fn package_path() -> &'static Path {
        Path::new("info/link.json")
    }

    fn from_str(str: &str) -> Result<Self, std::io::Error> {
        serde_json::from_str(str).map_err(Into::into)
    }
}

#[cfg(test)]
mod test {
    use super::LinkJson;
    use rstest::rstest;

    #[rstest]
    #[case::jupyterlab("link-json/jupyterlab-link.json")]
    #[case::setuptools("link-json/setuptools-link.json")]
    #[case::tzdata("link-json/tzdata-link.json")]
    fn test_link_json(#[case] path: &str) {
        let test_file = &crate::get_test_data_dir().join(path);
        let link_json: LinkJson =
            serde_json::from_reader(std::fs::File::open(test_file).unwrap()).unwrap();
        insta::assert_yaml_snapshot!(path, link_json);
    }
}
