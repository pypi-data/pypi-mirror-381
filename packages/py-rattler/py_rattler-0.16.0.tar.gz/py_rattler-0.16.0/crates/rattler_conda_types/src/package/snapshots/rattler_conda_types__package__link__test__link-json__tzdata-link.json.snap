---
source: crates/rattler_conda_types/src/package/link.rs
expression: link_json
---
noarch:
  type: generic
package_metadata_version: 1

